localstack/services/awslambda/invocation/version_manager.py

```diff
@@ -0,0 +1,381 @@
+import dataclasses
+import logging
+import queue
+import threading
+import uuid
+from concurrent.futures import Future
+from datetime import datetime
+from enum import Enum, auto
+from queue import Queue
+from threading import Thread
+from typing import Dict, List, Optional, Union
+
+from localstack.services.awslambda.invocation.lambda_models import (
+    FunctionVersion,
+    Invocation,
+    InvocationError,
+    InvocationLogs,
+    InvocationResult,
+    ServiceEndpoint,
+)
+from localstack.services.awslambda.invocation.runtime_environment import (
+    InvalidStatusException,
+    RuntimeEnvironment,
+    RuntimeStatus,
+)
+from localstack.services.awslambda.invocation.runtime_executor import (
+    cleanup_version,
+    prepare_version,
+)
+from localstack.utils.cloudwatch.cloudwatch_util import store_cloudwatch_logs
+
+LOG = logging.getLogger(__name__)
+
+
+class ValueNameEnum(Enum):
+    def __str__(self):
+        return str(self.name)
+
+
+class State(ValueNameEnum):
+    Pending = auto()
+    Active = auto()
+    Inactive = auto()
+    Failed = auto()
+
+
+class StateReasonCode(ValueNameEnum):
+    Idle = auto()
+    Creating = auto()
+    Restoring = auto()
+    InsufficientRolePermissions = auto()
+    InvalidConfiguration = auto()
+    InternalError = auto()
+    InvalidSecurityGroup = auto()
+    ImageDeleted = auto()
+    ImageAccessDenied = auto()
+    InvalidImage = auto()
+
+
+@dataclasses.dataclass(frozen=True)
+class VersionState:
+    state: State
+    code: Optional[StateReasonCode] = None
+    reason: Optional[str] = None
+
+
+# InvocationResultFuture = Future[InvocationResult]
+#
+# from typing_extensions import Future
+
+
+@dataclasses.dataclass(frozen=True)
+class QueuedInvocation:
+    invocation_id: str
+    result_future: "Future[InvocationResult]"
+    retries: int
+    invocation: Invocation
+
+
+@dataclasses.dataclass
+class RunningInvocation:
+    invocation: QueuedInvocation
+    start_time: datetime
+    executor: RuntimeEnvironment
+    logs: Optional[str] = None
+
+
+@dataclasses.dataclass(frozen=True)
+class LogItem:
+    log_group: str
+    log_stream: str
+    logs: str
+
+
+class ShutdownPill:
+    pass
+
+
+class LogHandler:
+    log_queue: "Queue[Union[LogItem, ShutdownPill]]"
+    _thread: Optional[Thread]
+    _shutdown_event: threading.Event
+
+    def __init__(self) -> None:
+        self.log_queue = Queue()
+        self._shutdown_event = threading.Event()
+        self._thread = None
+
+    def run_log_loop(self) -> None:
+        while not self._shutdown_event.is_set():
+            log_item = self.log_queue.get()
+            if isinstance(log_item, ShutdownPill):
+                return
```

Comment:
i think it's better to have a singleton object `QUEUE_SHUTDOWN = ShutdownPill()` and check for identity with `if log_item is QUEUE_SHUTDOWN` (it's a lot faster). better yet, you could just have `QUEUE_SHUTDOWN` be a `LogItem`, then we don't need the type union.

edit: just saw you're using `ShutdownPill` for other queues as well. so disregard the second sentence :-)